src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Table;
+using Newtonsoft.Json;
+using Orleans.Transactions.Abstractions;
+
+namespace Orleans.Transactions.AzureStorage
+{
+    public class AzureTableTransactionalStateStorage<TState> : ITransactionalStateStorage<TState>
+        where TState : class, new()
+    {
+        private readonly CloudTable table;
+        private readonly string partition;
+        JsonSerializerSettings JsonSettings;
+        private readonly ILogger logger;
+
+        private KeyEntity key;
+        private List<StateEntity> states;
+
+        public AzureTableTransactionalStateStorage(CloudTable table, string partition, JsonSerializerSettings JsonSettings, ILogger<AzureTableTransactionalStateStorage<TState>> logger)
+        {
+            this.table = table;
+            this.partition = partition;
+            this.JsonSettings = JsonSettings;
+            this.logger = logger;
+        }
+
+        public async Task<TransactionalStorageLoadResponse<TState>> Load()
+        {
+            try
+            {
+                Task<KeyEntity> keyTask = ReadKey();
+                Task<List<StateEntity>> statesTask = ReadStates();
+                this.key = await keyTask;
+                this.states = await statesTask;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                {
+                    return new TransactionalStorageLoadResponse<TState>();
+                }
+                TState commitedState = TryFindState(this.key.CommittedTransactionId, out TState found) ? found : new TState();
+                var pendingStates = states.Select(s => new PendingTransactionState<TState>(s.TransactionId, s.SequenceId, s.GetState<TState>(this.JsonSettings))).ToList();
```

Comment:
LINQ usage here is expressive but inefficient, particularly since we know the size of the result and therefore don't need a bunch of allocations for list resizing and for the Select closure. The alternative is to pre-allocate the pendingStates as `new List<PendingTransactionState>(states.Count)` and `foreach` over `states` adding them to the list.

Feel free to ignore, it's probably not a big deal

Reply:
I agree with the concern. In general link statements are almost never efficient.  I'm unconvinced, at this time, that the performance implications justify the maintainabilty cost of not using link.  This seems like premature optimization, do you disagree?